hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/CustomFractionalFeeAssessor.java

```diff
@@ -0,0 +1,206 @@
+                unitsLeft -= assessedAmount;
+                validateTrue(unitsLeft >= 0, INSUFFICIENT_SENDER_ACCOUNT_BALANCE_FOR_CUSTOM_FEE);
+
+                // make credit to the collector
+                final var map = result.getInputTokenAdjustments().getOrDefault(denom, new HashMap<>());
+                map.merge(collector, assessedAmount, Long::sum);
+                result.getInputTokenAdjustments().put(denom, map);
+
+                final var finalEffPayerNums =
+                        (filteredCredits == creditsForToken) ? effectivePayerAccounts : filteredCredits.keySet();
+                final var finalEffPayerNumsArray = new AccountID[finalEffPayerNums.size()];
+
+                // Add assessed custom fees to the result. This is needed to build transaction record
+                result.addAssessedCustomFee(AssessedCustomFee.newBuilder()
+                        .effectivePayerAccountId(finalEffPayerNums.toArray(finalEffPayerNumsArray))
+                        .feeCollectorAccountId(collector)
+                        .tokenId(denom)
+                        .amount(assessedAmount)
+                        .build());
+            }
+        }
+    }
+
+    private void adjustInputTokenTransfersWithReclaimAmounts(
+            final Map<TokenID, Map<AccountID, Long>> inputTokenAdjustments,
+            final TokenID denom,
+            final Map<AccountID, Long> filteredCredits) {
+        // if we reached here it means there are credits for the token
+        final var map = inputTokenAdjustments.get(denom);
+        for (final var entry : filteredCredits.entrySet()) {
+            final var account = entry.getKey();
+            final var amount = entry.getValue();
+            map.put(account, amount);
+        }
+        inputTokenAdjustments.put(denom, map);
+    }
+
+    private Map<AccountID, Long> filteredByExemptions(
+            final Map<AccountID, Long> creditsForToken, final CustomFeeMeta feeMeta, final CustomFee fee) {
+        final var filteredCredits = new HashMap<AccountID, Long>();
+        for (final var entry : creditsForToken.entrySet()) {
+            final var account = entry.getKey();
+            final var amount = entry.getValue();
+            if (!isPayerExempt(feeMeta, fee, account)) {
+                filteredCredits.put(account, amount);
+            }
+        }
+        return !filteredCredits.isEmpty() ? filteredCredits : creditsForToken;
+    }
+
+    private long amountOwed(final long initialUnits, @NonNull final FractionalFee fractionalFee) {
+        final var numerator = fractionalFee.fractionalAmount().numerator();
+        final var denominator = fractionalFee.fractionalAmount().denominator();
+        var nominalFee = 0L;
+        try {
+            nominalFee = safeFractionMultiply(numerator, denominator, initialUnits);
+        } catch (final ArithmeticException e) {
+            throw new HandleException(CUSTOM_FEE_OUTSIDE_NUMERIC_RANGE);
```

Comment:
It would be helpful to `log.error` the numerator, denominator, and initial units here

Reply:
I don't think this is an internal error of any sort, the `ArithmeticException` is basically for control flow here.

Reply:
Throwing exceptions is an expensive control flow mechanism, especially for crypto transfer, but I guess we don't have any data that shows this is a problem yet